invenio_sequencegenerator/models.py

```diff
@@ -0,0 +1,157 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Invenio.
+# Copyright (C) 2016 CERN.
+#
+# Invenio is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License as
+# published by the Free Software Foundation; either version 2 of the
+# License, or (at your option) any later version.
+#
+# Invenio is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Invenio; if not, write to the Free Software Foundation, Inc.,
+# 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+
+"""Models for storing counters for each pattern."""
+
+import re
+
+from invenio_db import db
+from sqlalchemy.orm.exc import MultipleResultsFound
+
+from .errors import InvalidCall, InvalidTemplate, NoSuchSequence
+from .utils import double_counter, extract_placeholders
+
+counter_regex = re.compile(r'({counter(!.)?(:.*)?})')
+
+
+class Sequence(db.Model):
+    """Represents sequence definitions."""
+
+    __tablename__ = 'sequencegenerator_sequence'
+
+    name = db.Column(db.String, primary_key=True)
+    """The identifying name of the sequence."""
+
+    template = db.Column(db.String, unique=True)
+    """The template of the sequence."""
+
+    parent = db.Column(db.ForeignKey(name))
+    """Indicate that the template depends on another one."""
+
+    @classmethod
+    def create(cls, name, template):
+        """Create a new sequence."""
+        placeholders = extract_placeholders(template)
+
+        # Check that {counter} exists and double its brackets
+        if not any(counter_regex.search('{{{0}}}'.format(p))
+                   for p in placeholders):
+            raise InvalidTemplate('No counter placeholder')
+        template = double_counter(template, counter_regex)
+
+        with db.session.begin_nested():
+            # Check if parent sequence exists
+            try:
+                parent = Sequence.query.filter(
+                    Sequence.name.in_(placeholders)
+                ).one_or_none()
+            except MultipleResultsFound:
+                raise InvalidTemplate('More than 1 parents in template')
+
+            seq = cls(
+                name=name,
+                template=template,
+                parent=parent.name if parent else None
+            )
+            db.session.add(seq)
+
+            return seq
+
+    @classmethod
+    def get(cls, name):
+        """Get the sequence with the given name."""
+        with db.session.begin_nested():
+            seq = db.session.query(cls).get(name)
+            if seq is None:
+                raise NoSuchSequence()
+            return seq
+
+    def next(self, **kwargs):
+        """Get the next report number for the given sequence."""
+        try:
```

Comment:
Consider creating `def _format(self, **kwargs)` method to de-duplicate code in `next` and `reset`.